src/collections.rs

```diff
@@ -0,0 +1,236 @@
+use async_std::channel::{Receiver, Sender};
+use async_std::sync::Mutex;
+use std::collections::VecDeque;
+use std::ops::Deref;
+use tracing::error;
+
+/// Blocking async double-ended queue with interior mutability.
+pub struct BlockingDeque<T> {
+    queue: Mutex<VecDeque<T>>,
+    control: (Sender<()>, Receiver<()>),
+}
+
+impl<T> BlockingDeque<T> {
+    /// Creates a new queue.
+    ///
+    /// The created queue has space to hold at most `cap` records at a time.
+    ///
+    /// # Panics
+    ///
+    /// Capacity must be a positive number. If `cap` is zero, this function will panic.
+    pub fn new(cap: usize) -> Self {
```

Comment:
I think this is slightly misleading because this is the size of the control, not the size of the inner queue.  Also, there is no way to configure the inner vedequee.  Can we change to `with_control_size(cap: usize)`?